File: buildBDD.m
function [pyvar,pyformula] = buildBDD(model,dels)
    nrxn=size(model.rxns,1);
    ngen=size(model.genes,1);
    var="'"+model.genes{1}+"'";
    for i=2:ngen
        var=var+", '"+model.genes{i}+"'";
    end
    pyvar="bdd.declare("+var+")";
    if dels(1)==1
        formula="("+model.grs{1}+")";
    elseif dels(1)==0
        formula="! ("+model.grs{1}+")";
    end
    for j=2:nrxn
        if isempty(model.grs{j})
            continue;
        end
        if dels(1)==1
            formula=formula+" & ("+model.grs{j}+")";
        elseif dels(1)==0
            formula=formula+" & ! ("+model.grs{j}+")";
        end
    end
    pyformula="u = bdd.add_expr(r'"+formula+"')";
end

%{
import dd.autoref as _bdd
bdd = _bdd.BDD()
bdd.declare('x', 'y', 'z')
u = bdd.add_expr()
assignment = bdd.pick(u)
%}

File: buildBDD.py
import dd.autoref as _bdd
from pathlib import Path
from cobra.io import read_sbml_model
import logging
import time

mini_fbc2_path = "e_coli_core.xml"
model = read_sbml_model(mini_fbc2_path)

num_reactions = len(model.reactions)
num_genes = len(model.genes)
bdd = _bdd.BDD()
for i in range(num_genes):
    bdd.add_var(model.genes[i].id)

grRules = []
for j in range(num_reactions):
    if model.reactions[j].gene_reaction_rule == "":
        continue
    grRule = model.reactions[j].gene_reaction_rule
    grRules.append("(" + grRule.replace('and', '&').replace('or', '|') + ")")

ultraFormula = " & ".join(grRules)
print(ultraFormula)

s = time.time()
u = bdd.add_expr(ultraFormula)
bdd.pick(u)
e = time.time()
print(f"{e-s} seconds")

print("sd")